### PR TITLE
[Tool] Generate input settings of script tool in meta info

### PR DESCRIPTION
**Pull request by lalala123123**
# Description
promptflow sdk supports the ability to config the input in @tool, likes below:
```
@tool(
    description="tool with input settings",
    input_settings={
        "teacher_id": InputSetting(enabled_by="user_type", enabled_by_value=[UserType.TEACHER]),
        "student_id": InputSetting(enabled_by="user_type", enabled_by_value=[UserType.STUDENT],
                                   undefined_field={"key": "value"}),
    },
    unknown_key="value",
)
    pass
```
However, only the meta generated by the tool in the package tool contains the configuration info. 

In this PR, the script tool generate meta will use the same logic as in the package tool. The meta generated by script tool will contain the input configuration.

Expect tools meta generated by tool script:
```json
{
    "package": {},
    "code": {
        "tool_with_input_settings": {
            "type": "python",
            "inputs": {
                "user_type": {"type": ["string"], "enum": ["student", "teacher"]},
                "student_id": {
                    "type": ["string"],
                    "enabled_by": "user_type",
                    "enabled_by_value": ["student"],
                    "undefined_field": {"key": "value"},
                },
                "teacher_id": {"type": ["string"], "enabled_by": "user_type", "enabled_by_value": ["teacher"]},
            },
            "description": "tool with input settings",
            "source": "tool_with_input_settings.py",
            "function": "tool_with_input_settings",
            "unknown_key": "value",
        }
    }
}
```
Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 1]
```diff
@@ -239,21 +239,28 @@ def collect_tool_function_in_module(m):
...
+    construct_tool, validate_result = tool_operations._serialize_tool(tool, input_settings, extra_info, f)
+    validate_result.try_raise(raise_error=True)
+    # Handler string enum in tool dict
+    construct_tool = json.loads(json.dumps(construct_tool))
+    return construct_tool
 
 
 def generate_python_meta_dict(name, content, source=None):
-    return asdict_without_none(generate_python_tool(name, content, source))
+    return generate_python_tool(name, content, source)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:263]: why orignially need asdict_without_none ?

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 2]
```diff
@@ -239,21 +239,28 @@ def collect_tool_function_in_module(m):
...
         tool.source = source
-    return tool
+    construct_tool, validate_result = tool_operations._serialize_tool(tool, input_settings, extra_info, f)
+    validate_result.try_raise(raise_error=True)
+    # Handler string enum in tool dict
+    construct_tool = json.loads(json.dumps(construct_tool))
+    return construct_tool
 
 
 def generate_python_meta_dict(name, content, source=None):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:262]: shall we remove this function? as generate_python_tool does exactly same thing

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 3]
```diff
@@ -239,21 +239,28 @@ def collect_tool_function_in_module(m):
...
     if name is not None:
         tool.name = name
     if source is None:
         tool.code = content
     else:
         tool.source = source
-    return tool
+    construct_tool, validate_result = tool_operations._serialize_tool(tool, input_settings, extra_info, f)
+    validate_result.try_raise(raise_error=True)
+    # Handler string enum in tool dict
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:257]: Handler -> Handle

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 2 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: fix it, remove this function

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 3 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: fix typo

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 1 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: generate_python_tool will also remove the none value, before returning the meta dict.
To simplify the meta dict or the requirements of extension, @zhengfeiwang can help confirm it?

**Review by wangchao1230 — APPROVED** [review]

**Review by crazygao — APPROVED** [review]

**Review by lumoslnt — APPROVED** [review]

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 4]
```diff
@@ -161,6 +166,237 @@ def _parse_tool_from_function(f, initialize_inputs=None, gen_custom_type_conn=Fa
...
+    :type func_path: str
+    :return: Validation result of the tool
+    :rtype: ValidationResult
+    """
+    from promptflow._sdk.entities._validation import ValidationResultBuilder
+
+    validate_result = ValidationResultBuilder.success()
+
+    if extra_info:
+        if ICON in extra_info:
```
Comment (thy09) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:191]: Such icon logic is also valid for script tool?

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 5]
```diff
@@ -238,27 +266,32 @@ def collect_tool_function_in_module(m):
         return tool_methods[0]
 
 
-def generate_python_tool(name, content, source=None):
+def generate_python_tool_meta_dict(name, content, source=None):
     m = load_python_module(content, source)
     f, initialize_inputs = collect_tool_function_in_module(m)
-    tool = _parse_tool_from_function(f, initialize_inputs=initialize_inputs)
+    tool = _parse_tool_from_function(
+        f, initialize_inputs=initialize_inputs, gen_custom_type_conn=True, skip_prompt_template=True
```
Comment (lumoslnt) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:273]: Why set the `gen_custom_type_conn` and `skip_prompt_template` to true? To my understand, these should be false for script tool.

Comment (lalala123123) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: revert it

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 6]
```diff
@@ -238,27 +266,30 @@ def collect_tool_function_in_module(m):
...
     if source is None:
         tool.code = content
     else:
         tool.source = source
-    return tool
-
-
-def generate_python_meta_dict(name, content, source=None):
-    return asdict_without_none(generate_python_tool(name, content, source))
+    construct_tool, validate_result = _serialize_tool(tool, input_settings, extra_info)
```
Comment (huaiyan) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:282]: Rename `validate_result` to `is_invlid_result`?
(validate is a verb

File: src/promptflow/promptflow/_core/tool_settings_parser.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                tool_inputs[input_name][k] = v
+        if settings.generated_by:
+            generated_by_inputs.update(settings.generated_by._input_settings)
+    tool_inputs.update(generated_by_inputs)
+
+
+def _parser_tool_icon(extra_info):
+    """parser tool icon to base64."""
+    if ICON in extra_info:
+        extra_info[ICON] = _serialize_icon_data(extra_info["icon"])
```
Comment (huaiyan) [inline src/promptflow/promptflow/_core/tool_settings_parser.py:45]:
```suggestion
        extra_info[ICON] = _serialize_icon_data(extra_info[ICON])
```

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 6 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: refine variable name

File: src/promptflow/promptflow/_core/tool_settings_parser.py (hunk 7 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_core/tool_settings_parser.py]: fix it

**Review by thy09 — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**